 # <center> Data Preprocessing dengan Pandas

## <div class="alert alert-block alert-info">Pengenalan Library Pandas</div>

![Pandas Icon](img/pandas.png)
Pandas adalah alat analisis dan manipulasi data open source yang cepat, kuat, fleksibel dan mudah digunakan, dibangun di atas bahasa pemrograman Python. Secara khusus, Pandas menawarkan operasi struktur data dan operasi untuk memanipulasi tabel numerik serta deret waktu (*time-series*)

## <div class="alert alert-block alert-info"> Operasi Dasar Pandas pada Dataframe </div>

### <div class="alert alert-block alert-success"> Import library </div>

In [2]:
# import library
import pandas as pd

pd adalah alias yang umum digunakan untuk library pandas. Hal ini digunakan untuk memudahkan penggunaan library pandas

In [4]:
# read dataset
df = pd.read_csv('../dataset/occupation.csv')

In [3]:
df.head(3)

user_id  age gender  occupation zip_code
0     1001   24      M  technician    85711
1     1002   53      F       other    94043
2     1003   23      M      writer    32067

<div class="alert alert-block alert-warning">

*read_csv* merupakan function dari library pandas. Untuk melihat parameter yang bisa digunakan, Anda bisa melihatnya dengan cara menjalankan *pd.read_csv?*

*head* merupakan method dari objek df. Sama seperti function, untuk melihat dokumentasinya Anda bisa mengakses dengan cara menjalankan *df.head?*
</div>

Pada dataset df, terdapat 5 kolom/atribut, yaitu user_id, age, gender, occupation, zip_code. Untuk mencetak nama-nama kolom yang ada, Anda dapat menjalankan code *df.columns*.

Dataset df memiliki 943 objek/poin/instance. Untuk mengecek dimensi dataset, Anda dapat menjalankan code *df.shape*. Nilai pertama adalah jumlah baris/row dan nilai kedua adalah jumlah kolom/column/atribut.

In [4]:
df.columns

Index(['user_id', 'age', 'gender', 'occupation', 'zip_code'], dtype='object')

In [5]:
df.shape

(943, 5)

### <div class="alert alert-block alert-success"> Manipulasi dataset </div>

**Merubah nama kolom**

In [6]:
df.rename(columns={'occupation':'job'}, inplace=True)

In [7]:
df

user_id  age gender            job zip_code
0       1001   24      M     technician    85711
1       1002   53      F          other    94043
2       1003   23      M         writer    32067
3       1004   24      M     technician    43537
4       1005   33      F          other    15213
..       ...  ...    ...            ...      ...
938     1939   26      F        student    33319
939     1940   32      M  administrator    02215
940     1941   20      M        student    97229
941     1942   48      F      librarian    78209
942     1943   22      M        student    77841

[943 rows x 5 columns]

**Slice dataset pada 1 kolom tertentu**

In [8]:
# akses hanya kolom age saja
df['age']

0      24
1      53
2      23
3      24
4      33
       ..
938    26
939    32
940    20
941    48
942    22
Name: age, Length: 943, dtype: int64

**Slice dataset lebih dari 1 kolom tertentu**

In [9]:
df[['age','gender']]

age gender
0     24      M
1     53      F
2     23      M
3     24      M
4     33      F
..   ...    ...
938   26      F
939   32      M
940   20      M
941   48      F
942   22      M

[943 rows x 2 columns]

**Akses hanya nilai dalam kolom**

In [10]:
df['age'].values

array([24, 53, 23, 24, 33, 42, 57, 36, 29, 53, 39, 28, 47, 45, 49, 21, 30,
       35, 40, 42, 26, 25, 30, 21, 39, 49, 40, 32, 41,  7, 24, 28, 23, 38,
       20, 19, 23, 28, 41, 38, 33, 30, 29, 26, 29, 27, 53, 45, 23, 21, 28,
       18, 26, 22, 37, 25, 16, 27, 49, 50, 36, 27, 31, 32, 51, 23, 17, 19,
       24, 27, 39, 48, 24, 39, 24, 20, 30, 26, 39, 34, 21, 50, 40, 32, 51,
       26, 47, 49, 43, 60, 55, 32, 48, 26, 31, 25, 43, 49, 20, 36, 15, 38,
       26, 27, 24, 61, 39, 44, 29, 19, 57, 30, 47, 27, 31, 40, 20, 21, 32,
       47, 54, 32, 48, 34, 30, 28, 33, 24, 36, 20, 59, 24, 53, 31, 23, 51,
       50, 46, 20, 30, 49, 13, 42, 53, 31, 45, 40, 33, 35, 20, 38, 33, 25,
       25, 32, 25, 57, 50, 23, 27, 50, 25, 49, 47, 20, 47, 37, 48, 52, 53,
       48, 55, 56, 30, 26, 28, 20, 26, 15, 22, 26, 36, 33, 37, 53, 39, 26,
       42, 32, 30, 33, 42, 29, 38, 42, 49, 55, 21, 30, 40, 27, 41, 25, 52,
       47, 14, 39, 43, 33, 39, 66, 49, 33, 26, 35, 22, 22, 37, 32, 30, 19,
       29, 19, 31, 51, 28

**Slice instance tertentu dengan kondisi dari nilai kolom**

In [11]:
# slice instance dengan gender male (M). 1 condition
df[df['gender']=='M']

user_id  age gender            job zip_code
0       1001   24      M     technician    85711
2       1003   23      M         writer    32067
3       1004   24      M     technician    43537
5       1006   42      M      executive    98101
6       1007   57      M  administrator    91344
..       ...  ...    ...            ...      ...
935     1936   24      M          other    32789
936     1937   48      M       educator    98072
939     1940   32      M  administrator    02215
940     1941   20      M        student    97229
942     1943   22      M        student    77841

[669 rows x 5 columns]

In [12]:
# slice instance dengan gender male (M) DAN berumur lebih dari 30 tahun. 2 condition
df[(df['gender']=='M') & (df['age']>30)]

user_id  age gender            job zip_code
5       1006   42      M      executive    98101
6       1007   57      M  administrator    91344
7       1008   36      M  administrator    05201
9       1010   53      M         lawyer    90703
12      1013   47      M       educator    29206
..       ...  ...    ...            ...      ...
931     1932   58      M       educator    06437
933     1934   61      M       engineer    22902
934     1935   42      M         doctor    66221
936     1937   48      M       educator    98072
939     1940   32      M  administrator    02215

[350 rows x 5 columns]

In [13]:
# slice instance dengan gender male (M) ATAU berumur lebih dari 30 tahun. 2 condition
df[(df['gender']=='M') | (df['age']>30)]

user_id  age gender            job zip_code
0       1001   24      M     technician    85711
1       1002   53      F          other    94043
2       1003   23      M         writer    32067
3       1004   24      M     technician    43537
4       1005   33      F          other    15213
..       ...  ...    ...            ...      ...
937     1938   38      F     technician    55038
939     1940   32      M  administrator    02215
940     1941   20      M        student    97229
941     1942   48      F      librarian    78209
942     1943   22      M        student    77841

[814 rows x 5 columns]

Menampilkan hanya nilai kolom job dengan kondisi pada 2 kolom, yaitu gender dan age

In [14]:
df[(df['gender']=='M') & (df['age']>60)]['job'].values

array(['retired', 'salesman', 'retired', 'retired', 'educator',
       'engineer', 'other', 'retired', 'healthcare', 'executive',
       'retired', 'retired', 'librarian', 'retired', 'engineer',
       'programmer', 'administrator', 'doctor', 'educator', 'engineer'],
      dtype=object)

### <div class="alert alert-block alert-danger"> Soal Latihan </div>

1. Tampunglah semua list pekerjaan ke dalam variabel 'job'

In [15]:
job = df['job'].values

2. Berapakah pelajar perempuan yang berusia 18 tahun atau lebih pada dataset tersebut?

In [16]:
df[(df['age']>18)&(df['gender']=='F')]

user_id  age gender            job zip_code
1       1002   53      F          other    94043
4       1005   33      F          other    15213
10      1011   39      F          other    30329
11      1012   28      F          other    06405
14      1015   49      F       educator    97301
..       ...  ...    ...            ...      ...
921     1922   29      F  administrator    21114
929     1930   28      F      scientist    07310
937     1938   38      F     technician    55038
938     1939   26      F        student    33319
941     1942   48      F      librarian    78209

[249 rows x 5 columns]

## <div class="alert alert-block alert-info"> Data Preprocessing </div>

In [5]:
df = pd.read_csv('../dataset/occupation.csv')

### <div class="alert alert-block alert-success"> Tipe atribut </div>

    1. user_id     :  nominal
    2. age         :  ratio
    3. gender      :  nominal
    4. occuppation :  nominal
    5. zip_code    :  ordinal

### <div class="alert alert-block alert-success"> Deskripsi statistik </div>

In [18]:
df.describe()

user_id         age
count   943.000000  943.000000
mean   1472.000000   34.051962
std     272.364951   12.192740
min    1001.000000    7.000000
25%    1236.500000   25.000000
50%    1472.000000   31.000000
75%    1707.500000   43.000000
max    1943.000000   73.000000

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      942 non-null    object
 3   occupation  942 non-null    object
 4   zip_code    942 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


### <div class="alert alert-block alert-success"> Pengecekan problem pada dataset </div>

**Missing Value**

In [20]:
# cek missing value
df.isnull().sum()

user_id       0
age           0
gender        1
occupation    1
zip_code      1
dtype: int64

**Pengecekan instance yang memiliki missing value**

In [21]:
df[df.isnull().any(axis=1)]

user_id  age gender     occupation zip_code
24     1025   39      M            NaN    55107
73     1074   39    NaN      scientist    T8H1N
79     1080   34      F  administrator      NaN

**Penanganan missing value dengan menghapus missing value**

In [22]:
df = df.dropna()

In [23]:
df.isnull().sum()

user_id       0
age           0
gender        0
occupation    0
zip_code      0
dtype: int64

### <div class="alert alert-block alert-danger"> Soal Latihan </div>

1. Kenapa missing value perlu diidentifikasi dan ditangani?

Sebelum data dapat diolah, kualitas data harus dipastikan agar dapat memberikan informasi yang valid. Salah satu komponen kunci kualitas data adalah kelengkapan (*completeness*) data. Missing value berarti ada data objek yang tidak memiliki salah satu atau lebih atribut. Jika ada missing value dalam data, maka komponen kelengkapan data tidak dapat dipenuhi sehingga dapat mengganggu proses analisis data secara teknis serta dapat memberikan informasi yang tidak valid dan menyeluruh.

2. Selain dengan cara menghapuskan instance yang memiliki missing value, apakah cara penanganan lain yang dapat dilakukan untuk mengatasi permasalahan missing value? Jelaskan dan sebutkan !

Missing value dapat ditangani dengan teknik berikut :
    1. Mengabaikan data objek. Data objek/instance yang memiliki missing value tidak dihapus dari dataset, namun diabaikan dan tidak diikutkan dalam analisis data.
    2. Memperkirakan nilai missing value. Nilai missing value dapat diperkirakan dengan berbagai cara. Diantaranya, mengisi re

## <div class="alert alert-block alert-info"> Ekplorasi Data </div>

Untuk ekplorasi data, terlebih dahulu Anda harus merumuskan pertanyaan yang ingin dicari dalam dataset. Misalnya, *berapa jumlah pekerjaan yang ada?*, *berapa rata-rata umur wanita yang bekerja sebagai dokter?*, *apa saja pekerjaan yang tercatat* dan sebagainya.

Merumuskan pertanyan-pertanyaan tersebut adalah syarat awal dalam membuat visualisasi. Dari pertanyaan tersebut, muncul informasi yang ingin kita sampaikan melalui visualisasi nantinya.

### <div class="alert alert-block alert-success"> Mencari nilai unik dan jumlahnya </div>

In [24]:
df['occupation'].unique()

array(['technician', 'other', 'writer', 'executive', 'administrator',
       'student', 'lawyer', 'educator', 'scientist', 'entertainment',
       'programmer', 'librarian', 'homemaker', 'artist', 'engineer',
       'marketing', 'none', 'healthcare', 'retired', 'salesman', 'doctor'],
      dtype=object)

In [25]:
df['occupation'].value_counts()

student          196
other            105
educator          95
administrator     78
programmer        66
engineer          66
librarian         51
writer            45
executive         32
scientist         30
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: occupation, dtype: int64

### <div class="alert alert-block alert-success"> Summarization </div>

Summarization atau ringkasan dari data bertujuan untuk mencari gambaran besar dari dataset.

Dalam Pandas, terdapat dua method yang dapat digunakan berkaitan dengan summarization. Aggregate berfungsi untuk melihat ringkasan data secara keseluruhan. Sedangkan *groupby()* berfungsi untuk melihat ringkasan dari subset data.

<div class="alert alert-block alert-warning">
    Method Aggregation Pandas
</div>

|Method|Deskripsi|
|:-|:-|
|*count()*|Total item|
|*min()*, *max()*|Nilai paling kecil dan paling besar|
|*first()*, *last()*|Nilai awal dan akhir|
|*mean()*, *median()*|Rerata dan nilai tengah|
|*std()*, *var()*|Standar deviasi dan variance|
|*prod()*|Dot product|
|*sum()*|Jumlah/sum dari item|

<div class="alert alert-block alert-warning">
</div>

In [26]:
df.max()

user_id         1943
age               73
gender             M
occupation    writer
zip_code       Y1A6B
dtype: object

In [27]:
df.min()

user_id                1001
age                       7
gender                    F
occupation    administrator
zip_code              00000
dtype: object

In [28]:
df.count()

user_id       940
age           940
gender        940
occupation    940
zip_code      940
dtype: int64

In [29]:
df.mean()

user_id    1473.315957
age          34.041489
dtype: float64

### <div class="alert alert-block alert-success"> Mencari informasi dalam sub-dataset </div>

Berikut merupakan contoh dari penggunaan method *groupby()*.

In [30]:
df.groupby('occupation').count()

user_id  age  gender  zip_code
occupation                                   
administrator       78   78      78        78
artist              28   28      28        28
doctor               7    7       7         7
educator            95   95      95        95
engineer            66   66      66        66
entertainment       18   18      18        18
executive           32   32      32        32
healthcare          16   16      16        16
homemaker            7    7       7         7
lawyer              12   12      12        12
librarian           51   51      51        51
marketing           26   26      26        26
none                 9    9       9         9
other              105  105     105       105
programmer          66   66      66        66
retired             14   14      14        14
salesman            12   12      12        12
scientist           30   30      30        30
student            196  196     196       196
technician          27   27      27        27
writer              45   45      45        45

In [31]:
df.groupby('occupation', as_index=False).count()

occupation  user_id  age  gender  zip_code
0   administrator       78   78      78        78
1          artist       28   28      28        28
2          doctor        7    7       7         7
3        educator       95   95      95        95
4        engineer       66   66      66        66
5   entertainment       18   18      18        18
6       executive       32   32      32        32
7      healthcare       16   16      16        16
8       homemaker        7    7       7         7
9          lawyer       12   12      12        12
10      librarian       51   51      51        51
11      marketing       26   26      26        26
12           none        9    9       9         9
13          other      105  105     105       105
14     programmer       66   66      66        66
15        retired       14   14      14        14
16       salesman       12   12      12        12
17      scientist       30   30      30        30
18        student      196  196     196       196
19     technician       27   27      27        27
20         writer       45   45      45        45

Untuk melihat ringkasan dari subset data kita dapat menambahkan method aggregate di belakang method *groupby()*

In [32]:
df.groupby('occupation', as_index=False).age.mean()

occupation        age
0   administrator  38.807692
1          artist  31.392857
2          doctor  43.571429
3        educator  42.010526
4        engineer  36.348485
5   entertainment  29.222222
6       executive  38.718750
7      healthcare  41.562500
8       homemaker  32.571429
9          lawyer  36.750000
10      librarian  40.000000
11      marketing  37.615385
12           none  26.555556
13          other  34.523810
14     programmer  33.121212
15        retired  63.071429
16       salesman  35.666667
17      scientist  35.433333
18        student  22.081633
19     technician  33.148148
20         writer  36.311111

In [33]:
df.groupby(['occupation', 'gender']).age.mean()

occupation     gender
administrator  F         40.828571
               M         37.162791
artist         F         30.307692
               M         32.333333
doctor         M         43.571429
educator       F         39.115385
               M         43.101449
engineer       F         29.500000
               M         36.562500
entertainment  F         31.000000
               M         29.000000
executive      F         44.000000
               M         38.172414
healthcare     F         39.818182
               M         45.400000
homemaker      F         34.166667
               M         23.000000
lawyer         F         39.500000
               M         36.200000
librarian      F         40.000000
               M         40.000000
marketing      F         37.200000
               M         37.875000
none           F         36.500000
               M         18.600000
other          F         35.472222
               M         34.028986
programmer     F         32.16666

In [34]:
df.groupby('occupation', as_index=False).age.agg(['min', 'max'])

min  max
occupation             
administrator   21   70
artist          19   48
doctor          28   64
educator        23   63
engineer        22   70
entertainment   15   50
executive       22   69
healthcare      22   62
homemaker       20   50
lawyer          21   53
librarian       23   69
marketing       24   55
none            11   55
other           13   64
programmer      20   63
retired         51   73
salesman        18   66
scientist       23   55
student          7   42
technician      21   55
writer          18   60

### <div class="alert alert-block alert-danger"> Soal Latihan </div>

1. Pekerjaan apa yang paling banyak dijalani oleh perempuan?

In [35]:
df[df['gender']=='F'].groupby('occupation', as_index=False).user_id.count().sort_values('user_id')

occupation  user_id
18     technician        1
14        retired        1
3        engineer        2
4   entertainment        2
8          lawyer        2
5       executive        3
16      scientist        3
15       salesman        3
11           none        4
7       homemaker        6
13     programmer        6
10      marketing       10
6      healthcare       11
1          artist       13
19         writer       19
2        educator       26
9       librarian       29
0   administrator       35
12          other       36
17        student       60

In [36]:
df.groupby(['gender','occupation'], as_index=False).user_id.count()

gender     occupation  user_id
0       F  administrator       35
1       F         artist       13
2       F       educator       26
3       F       engineer        2
4       F  entertainment        2
5       F      executive        3
6       F     healthcare       11
7       F      homemaker        6
8       F         lawyer        2
9       F      librarian       29
10      F      marketing       10
11      F           none        4
12      F          other       36
13      F     programmer        6
14      F        retired        1
15      F       salesman        3
16      F      scientist        3
17      F        student       60
18      F     technician        1
19      F         writer       19
20      M  administrator       43
21      M         artist       15
22      M         doctor        7
23      M       educator       69
24      M       engineer       64
25      M  entertainment       16
26      M      executive       29
27      M     healthcare        5
28      M      homemaker        1
29      M         lawyer       10
30      M      librarian       22
31      M      marketing       16
32      M           none        5
33      M          other       69
34      M     programmer       60
35      M        retired       13
36      M       salesman        9
37      M      scientist       27
38      M        student      136
39      M     technician       26
40      M         writer       26

2. Pekerjaan apa yang memiliki pekerja perempuan dan pria termuda?

In [37]:
df.groupby(['gender','occupation'], as_index=False).age.min().sort_values(['gender','age'])

gender     occupation  age
17      F        student   13
12      F          other   18
15      F       salesman   18
19      F         writer   18
1       F         artist   19
7       F      homemaker   20
0       F  administrator   22
6       F     healthcare   22
2       F       educator   23
3       F       engineer   23
4       F  entertainment   24
10      F      marketing   24
9       F      librarian   26
11      F           none   26
13      F     programmer   26
16      F      scientist   26
8       F         lawyer   28
18      F     technician   38
5       F      executive   39
14      F        retired   70
38      M        student    7
32      M           none   11
33      M          other   13
25      M  entertainment   15
21      M         artist   20
34      M     programmer   20
20      M  administrator   21
29      M         lawyer   21
36      M       salesman   21
39      M     technician   21
40      M         writer   21
24      M       engineer   22
26      M      executive   22
28      M      homemaker   23
30      M      librarian   23
37      M      scientist   23
31      M      marketing   24
23      M       educator   25
22      M         doctor   28
27      M     healthcare   28
35      M        retired   51

3. Berapa perbandingan guru wanita dan guru pria?

In [38]:
df.groupby(['gender','occupation']).user_id.count()

gender  occupation   
F       administrator     35
        artist            13
        educator          26
        engineer           2
        entertainment      2
        executive          3
        healthcare        11
        homemaker          6
        lawyer             2
        librarian         29
        marketing         10
        none               4
        other             36
        programmer         6
        retired            1
        salesman           3
        scientist          3
        student           60
        technician         1
        writer            19
M       administrator     43
        artist            15
        doctor             7
        educator          69
        engineer          64
        entertainment     16
        executive         29
        healthcare         5
        homemaker          1
        lawyer            10
        librarian         22
        marketing         16
        none               5
        other        

##  <center><div class="alert alert-block alert-danger"> TUGAS </div>

### <div class="alert alert-block alert-danger"> Read Dataset </div>

Baca dataset titanic, kemudian tampung di variabel dt

In [39]:
dt = pd.read_csv('../dataset/titanic.csv')

Tampilkan 10 instance paling pertama dalam dataset tersebut

In [40]:
dt.head(10)

Survived  Pclass                                               Name  \
0         0     3.0                             Mr. Owen Harris Braund   
1         1     1.0  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1     3.0                              Miss. Laina Heikkinen   
3         1     1.0        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0     3.0                            Mr. William Henry Allen   
5         0     3.0                                    Mr. James Moran   
6         0     1.0                             Mr. Timothy J McCarthy   
7         0     3.0                      Master. Gosta Leonard Palsson   
8         1     3.0   Mrs. Oscar W (Elisabeth Vilhelmina Berg) Johnson   
9         1     2.0                 Mrs. Nicholas (Adele Achem) Nasser   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500  
5    male  27.0                        0                        0   8.4583  
6    male  54.0                        0                        0  51.8625  
7    male   2.0                        3                        1  21.0750  
8  female  27.0                        0                        2  11.1333  
9  female  14.0                        1                        0  30.0708

Berapakah jumlah instance dalam dataset tersebut ?

In [41]:
dt.shape

(887, 8)

Pada dataset titanic terdapat data penumpang yang menaiki kapal Titanic pada tahun 1912. Dataset tersebut memiliki delapan atribut, yaitu :

1. Survive = penanda masih hidupnya penumpang setelah kecelakaan. 1 masih hidup, 0 meninggal
2. Pclass = passenger class atau kelas penumpang. 1 kelas bisnis, 2 kelas eksekutif, 3 kelas ekonomi
3. Name = nama dari penumpang
4. Sex = jenis kelamin dari penumpang
5. Siblings/Spouses Aboard = jumlah saudara atau pasangan dari penumpang yang ikut dalam Titanic
6. Parents/Children Aboard = jumlah orangtua atau anak dari penumpang yang ikut dalam Titanic
7. Fare = biaya yang dikeluarkan penumpang dalam nilai USD, (.) titik berarti desimal

Ubah nama kolom menjadi sebagai berikut :

    1. Survive = survive
    2. Pclass = class
    3. Name = name
    4. Sex = sex
    5. Siblings/Spouses Aboard = siblings-spouses
    6. Parents/Children Aboard = parents-childs
    7. Fare = fare
    8. Age = age

In [42]:
dt.rename(columns={'Survived':'survive', 'Pclass':'class', 'Name':'name', 'Sex':'sex', 'Age':'age', 'Siblings/Spouses Aboard':'siblings-spouses',
       'Parents/Children Aboard':'parents_childs', 'Fare':'fare'}, inplace=True)

Tentukan tipe masing-masing atribut

    1. survive = nominal
    2. class = ordinal
    3. name = nominal
    4. sex = nominal
    5. siblings-spouses = ratio
    6. parents-childs = ratio
    7. fare = ratio
    8. age = ratio

### <div class="alert alert-block alert-danger"> Preprocessing </div>

Tentukan apakah ada missing value

In [43]:
dt.isnull().sum()

survive             0
class               1
name                0
sex                 0
age                 0
siblings-spouses    0
parents_childs      0
fare                2
dtype: int64

Tampilkan instance yang memiliki missing value

In [44]:
dt[dt.isnull().any(axis=1)]

survive  class                        name     sex   age  \
22        1    3.0          Miss. Anna McGowan  female  15.0   
39        1    3.0  Miss. Jamila Nicola-Yarred  female  14.0   
60        1    NaN          Miss. Amelie Icard  female  38.0   

    siblings-spouses  parents_childs  fare  
22                 0               0   NaN  
39                 1               0   NaN  
60                 0               0  80.0

Tangani missing value tersebut dengan menghapus instance yang memiliki missing value

In [45]:
dt = dt.dropna()

Buktikan bahwa tidak ada lagi missing value pada dataset

In [46]:
dt.isnull().sum()

survive             0
class               0
name                0
sex                 0
age                 0
siblings-spouses    0
parents_childs      0
fare                0
dtype: int64

Tampilkan hanya atribut nama dan umur penumpang

In [47]:
dt[['name','age']]

name   age
0                               Mr. Owen Harris Braund  22.0
1    Mrs. John Bradley (Florence Briggs Thayer) Cum...  38.0
2                                Miss. Laina Heikkinen  26.0
3          Mrs. Jacques Heath (Lily May Peel) Futrelle  35.0
4                              Mr. William Henry Allen  35.0
..                                                 ...   ...
882                               Rev. Juozas Montvila  27.0
883                        Miss. Margaret Edith Graham  19.0
884                     Miss. Catherine Helen Johnston   7.0
885                               Mr. Karl Howell Behr  26.0
886                                 Mr. Patrick Dooley  32.0

[884 rows x 2 columns]

Tampung semua nama penumpang pada kelas Bisnis yang memiliki umur lebih dari 50 tahun ke dalam variabel name_B_50plus

In [48]:
name_B_50plus = dt[(dt['class']==3)&(dt['age']>50)]['name'].values

### <div class="alert alert-block alert-danger"> Data Eksplorasi </div>

1. Berapakah jumlah penumpang Bisnis?

In [49]:
dt[dt['class']==1].name.count()

215

Terdapat **215** penumpang kelas Bisnis

2. Berapakah jumlah penumpang wanita kelas Ekonomi?

In [50]:
dt[(dt['class']==3) & (dt['sex']=='female')].name.count()

142

Terdapat **142** penumpang wanita kelas Ekonomi

3. Berapakah harga tiket paling mahal yang harus dikeluarkan penumpang kelas Ekonomi?

In [51]:
dt[dt['class']==3].fare.max()

69.55

Tiket kelas Ekonomi paling mahal adalah sebesar **69.55 USD**

4. Berapakah jumlah penumpang pria pada kelas Ekonomi?

In [52]:
dt[(dt['class']==3) & (dt['sex']=='male')].name.count()

343

Terdapat **343** penumpang kelas Ekonomi

5. Kelas manakah yang memiliki paling banyak penumpang yang single (tidak mengajak pasangan)?

In [60]:
dt.groupby(['class','siblings-spouses'], as_index=False).name.count()

class  siblings-spouses  name
0     1.0                 0   136
1     1.0                 1    71
2     1.0                 2     5
3     1.0                 3     3
4     2.0                 0   120
5     2.0                 1    55
6     2.0                 2     8
7     2.0                 3     1
8     3.0                 0   346
9     3.0                 1    82
10    3.0                 2    15
11    3.0                 3    12
12    3.0                 4    18
13    3.0                 5     5
14    3.0                 8     7

Dari tabel di atas dapat dilihat bahwa **Kelas Ekonomi (3) memiliki paling banyak penumpang single, yaitu sejumlah 346 orang.**

6. Pada kelas Bisnis, gender manakah yang paling banyak single (tidak mengajak pasangan)?

In [54]:
dt.groupby(['class','siblings-spouses','sex'], as_index=False).name.count()

class  siblings-spouses     sex  name
0     1.0                 0  female    48
1     1.0                 0    male    88
2     1.0                 1  female    40
3     1.0                 1    male    31
4     1.0                 2  female     3
5     1.0                 2    male     2
6     1.0                 3  female     2
7     1.0                 3    male     1
8     2.0                 0  female    44
9     2.0                 0    male    76
10    2.0                 1  female    28
11    2.0                 1    male    27
12    2.0                 2  female     3
13    2.0                 2    male     5
14    2.0                 3  female     1
15    3.0                 0  female    80
16    3.0                 0    male   266
17    3.0                 1  female    37
18    3.0                 1    male    45
19    3.0                 2  female     7
20    3.0                 2    male     8
21    3.0                 3  female     8
22    3.0                 3    male     4
23    3.0                 4  female     6
24    3.0                 4    male    12
25    3.0                 5  female     1
26    3.0                 5    male     4
27    3.0                 8  female     3
28    3.0                 8    male     4

Dari tabel di atas, jumlah penumpang pada kelas Bisnis (1) didominasi oleh penumpang **Pria, yaitu sebanyak 88 orang.**